src/bin/inx-chronicle/api/stardust/poi/merkle_hasher.rs

```diff
@@ -1,54 +1,133 @@
-                let r = self.hash(&data[k..]);
-                self.hash_node(&l, &r)
+                let l = Self::hash(&data[..k]);
+                let r = Self::hash(&data[k..]);
+                Self::hash_node(l, r)
             }
         }
     }
 
-    pub fn hash_empty(&self) -> Output<H> {
-        H::digest([])
+    fn hash_empty() -> Output<Blake2b256> {
+        Blake2b256::digest([])
     }
 
-    pub fn hash_leaf(&self, l: impl AsRef<[u8]>) -> Output<H> {
-        let mut hasher = H::default();
+    /// Hash a terminating leaf of the tree.
+    pub fn hash_leaf(l: impl AsRef<[u8]>) -> Output<Blake2b256> {
+        let mut hasher = Blake2b256::default();
         hasher.update([LEAF_HASH_PREFIX]);
         hasher.update(l);
         hasher.finalize()
     }
 
-    pub fn hash_node(&self, l: impl AsRef<[u8]>, r: impl AsRef<[u8]>) -> Output<H> {
-        let mut hasher = H::default();
+    /// Hash a subtree.
+    pub fn hash_node(l: impl AsRef<[u8]>, r: impl AsRef<[u8]>) -> Output<Blake2b256> {
+        let mut hasher = Blake2b256::default();
         hasher.update([NODE_HASH_PREFIX]);
         hasher.update(l);
         hasher.update(r);
         hasher.finalize()
     }
+
+    /// Create a merkle proof given a list of block IDs and a chosen block ID. The chosen leaf will become a
+    /// value node, and the path will contain all hashes above it. The remaining branches will be terminated early.
+    pub fn create_proof(block_ids: &[BlockId], chosen_block_id: &BlockId) -> Result<MerkleProof, CreateProofError> {
+        let index = block_ids
+            .iter()
+            .position(|id| id == chosen_block_id)
+            .ok_or_else(|| CreateProofError::BlockNotIncluded(chosen_block_id.to_hex()))?;
+        Self::create_proof_from_index(block_ids, index)
+    }
+
+    // NOTE:
+    // * `block_ids` is the list of past-cone block ids in "White Flag" order;
+    // * `block_ids.len() >= 2` must be true, or this function panics;
+    // * `index < block_ids.len()` must be true, or this function panics;
+    fn create_proof_from_index(block_ids: &[BlockId], index: usize) -> Result<MerkleProof, CreateProofError> {
+        if block_ids.len() < 2 {
+            Err(CreateProofError::InsufficientBlockIds(block_ids.len()))
+        } else {
+            let n = block_ids.len();
+            debug_assert!(index < n);
+
+            let data = block_ids.iter().map(|block_id| block_id.0).collect::<Vec<_>>();
+            Ok(Self::compute_proof(&data, index))
+        }
+    }
+
+    /// Recursively compute a merkle tree.
+    fn compute_proof(data: &[[u8; BlockId::LENGTH]], index: usize) -> MerkleProof {
+        let n = data.len();
+        debug_assert!(index < n);
+        match n {
+            0 | 1 => unreachable!(),
+            // The terminating point, where we only have two values that become
+            // left and right leaves. The chosen index is a `Value` while
+            // the other is a `Node`.
```

Comment:
This is a bit of a weird argument. It took me several hours to figure out this code well enough to write these comments. You so strongly disagree with adding helpful information that has no effect on compilation or runtime?

Reply:
I am saying that describing how the algorithm works can be done in the `fn` docs as there's no single line in the algorithm that necessary to explain. It's just normal Rust, no `unsafe` code, no weird aliasing, no nothing that would need explaining. What I would do is write how the basic algorithm works in general in the `fn` docs and keep the logic itself lean.

Reply:
Yes, but what I'm saying is that it's NOT clear what it does in principal 😆 . Just because you aren't using `unsafe` does not mean it's clear. Recursive code in particular is always more difficult to reason about.